ontolearner/learner/taxonomy_discovery/sbunlp.py

```diff
@@ -0,0 +1,317 @@
+# Copyright (c) 2025 SciKnowOrg
+#
+# Licensed under the MIT License (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://opensource.org/licenses/MIT
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+import json
+import importlib.util
+from typing import Any, Dict, List, Optional, Tuple
+
+import torch
+from transformers import AutoTokenizer, AutoModelForCausalLM
+
+from ...base import AutoLearner
+
+class SBUNLPFewShotLearner(AutoLearner):
+    """
+    Taxonomy-discovery via N×M batch prompting with a small Qwen model.
+
+    Lifecycle
+    ---------
+    fit():
+        Cache + clean training parent–child pairs.
+    predict():
+        Chunk (train pairs × test terms), prompt per chunk pair, parse, merge,
+        and deduplicate predicted relations.
+    """
+
+    def __init__(
+        self,
+        model_name: str = "Qwen/Qwen2.5-0.5B-Instruct",
+        try_4bit: bool = True,
+        num_train_chunks: int = 7,
+        num_test_chunks: int = 7,
+        max_new_tokens: int = 140,
+        max_input_tokens: int = 1500,
+        temperature: float = 0.0,
+        top_p: float = 1.0,
+        limit_num_prompts: Optional[int] = None,
+        output_dir: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.model_name = model_name
+        self.try_4bit = try_4bit
+
+        self.num_train_chunks = num_train_chunks
+        self.num_test_chunks = num_test_chunks
+
+        self.max_new_tokens = max_new_tokens
+        self.max_input_tokens = max_input_tokens
+        self.temperature = temperature
+        self.top_p = top_p
+        self.limit_num_prompts = limit_num_prompts
+
+        self.output_dir = output_dir
+
+        self.tokenizer: Optional[AutoTokenizer] = None
+        self.model: Optional[AutoModelForCausalLM] = None
+        self.device = "cuda" if torch.cuda.is_available() else "cpu"
```

Comment:
The device should be something that passed by user in argument! and default value it would be great to passed as a cpu!